File: src/scanners/smart.cpp
/// @file scanners/smart.cpp
/// @brief Smart detection: boolean-blind SQLi, context-aware XSS, WAF
///        detection, wildcard filtering, deduplication support.
#include "scanner_base.hpp"
#include <functional>
#include <numeric>

namespace apex {
namespace {

/// Detect WAF presence by sending malicious payloads and checking responses.
bool detect_waf(HttpClient &http, const std::string &url) {
  auto resp = http.get(url + "?test=<script>alert(1)</script>' OR 1=1--");
  if (resp.status_code == 403 || resp.status_code == 406) return true;
  const std::vector<std::string> waf_sigs = {
      "cloudflare", "akamai", "incapsula", "sucuri", "wordfence",
      "mod_security", "blocked", "forbidden", "waf"};
  return contains_any(resp.body, waf_sigs);
}

/// Check if target returns wildcard responses (same for any input).
bool is_wildcard(HttpClient &http, const std::string &url,
                 const std::string &param) {
  auto r1 = http.get(url + "?" + param + "=randomvalue12345");
  auto r2 = http.get(url + "?" + param + "=differentvalue67890");
  if (r1.body.empty() || r2.body.empty()) return false;
  size_t diff = (r1.body.size() > r2.body.size())
                    ? r1.body.size() - r2.body.size()
                    : r2.body.size() - r1.body.size();
  return diff < 10 && r1.status_code == r2.status_code;
}

/// Boolean-blind SQL injection via response size differential.
std::vector<Finding> scan_sqli_blind_boolean(const Config &, HttpClient &http,
                                            const CrawlResult &crawl) {
  std::vector<Finding> findings;
  const std::string true_payload = "' AND '1'='1";
  const std::string false_payload = "' AND '1'='2";

  for (const auto &url : crawl.urls) {
    auto targets = get_targets(crawl, url);
    for (const auto &[base, param] : targets) {
      if (is_wildcard(http, url, param)) continue;
      auto true_resp = http.get(base + true_payload);
      auto false_resp = http.get(base + false_payload);
      size_t size_diff = (true_resp.body.size() > false_resp.body.size())
                             ? true_resp.body.size() - false_resp.body.size()
                             : false_resp.body.size() - true_resp.body.size();
      if (size_diff > 50 && true_resp.status_code == false_resp.status_code) {
        findings.push_back({"SQLi (Blind)", "high", url,
                            "Boolean-blind SQLi: size diff=" +
                                std::to_string(size_diff),
                            param, true_payload, ""});
      }
    }
  }
  return findings;
}

/// Detect XSS reflection context and use context-appropriate payloads.
std::vector<Finding> scan_xss_context_aware(const Config &, HttpClient &http,
                                            const CrawlResult &crawl) {
  std::vector<Finding> findings;
  const std::string canary = "apex7x7z";

  struct CtxPayload {
    const char *ctx;
    const char *payload;
  };
  const CtxPayload ctx_payloads[] = {
      {"html", "<img src=x onerror=alert(1)>"},
      {"attr", "\" onmouseover=alert(1) x=\""},
      {"js", "';alert(1);//"},
      {"url", "javascript:alert(1)"},
  };

  for (const auto &url : crawl.urls) {
    auto targets = get_targets(crawl, url, "q");
    for (const auto &[base, param] : targets) {
      auto resp = http.get(base + canary);
      if (resp.body.find(canary) == std::string::npos) continue;

      // Detect context.
      std::string ctx = "html";
      size_t pos = resp.body.find(canary);
      if (pos > 0) {
        std::string before = resp.body.substr(
            pos > 30 ? pos - 30 : 0, pos > 30 ? 30 : pos);
        if (before.find("=\"") != std::string::npos ||
            before.find("='") != std::string::npos)
          ctx = "attr";
        else if (before.find("var ") != std::string::npos ||
                 before.find("'") != std::string::npos)
          ctx = "js";
        else if (before.find("href=") != std::string::npos)
          ctx = "url";
      }

      for (const auto &cp : ctx_payloads) {
        if (ctx != cp.ctx) continue;
        auto test = http.get(base + std::string(cp.payload));
        if (test.body.find(cp.payload) != std::string::npos) {
          findings.push_back({"XSS (Context)", "high", url,
                              "Context-aware XSS in " + ctx + " context",
                              param, cp.payload, ""});
          break;
        }
      }
    }
  }
  return findings;
}

/// WAF detection and reporting.
std::vector<Finding> scan_waf_detect(const Config &, HttpClient &http,
                                     const CrawlResult &crawl) {
  std::vector<Finding> findings;
  if (crawl.urls.empty()) return findings;
  std::string base = base_url_from(crawl.urls[0]);

  struct WafSig {
    const char *header;
    const char *value;
    const char *name;
  };
  const WafSig sigs[] = {
      {"Server", "cloudflare", "Cloudflare"},
      {"X-Sucuri-ID", "", "Sucuri"},
      {"X-CDN", "Incapsula", "Imperva/Incapsula"},
      {"Server", "AkamaiGHost", "Akamai"},
      {"X-Powered-By", "wordfence", "Wordfence"},
  };

  auto resp = http.get(base + "/?test=<script>alert(1)</script>");
  for (const auto &sig : sigs) {
    auto it = resp.headers.find(sig.header);
    if (it != resp.headers.end()) {
      std::string needle = sig.value;
      if (needle.empty() || it->second.find(needle) != std::string::npos) {
        findings.push_back({"WAF Detected", "info", base,
                            std::string("WAF: ") + sig.name, "", "", ""});
      }
    }
  }

  if (resp.status_code == 403 || resp.status_code == 406) {
    findings.push_back({"WAF Detected", "info", base,
                        "WAF blocking malicious input (HTTP " +
                            std::to_string(resp.status_code) + ")",
                        "", "", ""});
  }
  return findings;
}

/// Time-based blind SQL injection.
std::vector<Finding> scan_sqli_time_based(const Config &, HttpClient &http,
                                          const CrawlResult &crawl) {
  std::vector<Finding> findings;
  const std::vector<std::string> payloads = {
      "' AND SLEEP(5)--", "'; WAITFOR DELAY '0:0:5'--",
      "' AND pg_sleep(5)--", "1; SELECT SLEEP(5)"};

  for (const auto &url : crawl.urls) {
    auto targets = get_targets(crawl, url);
    for (const auto &[base, param] : targets) {
      auto baseline = http.get(base + "test");
      for (const auto &payload : payloads) {
        auto resp = http.get(base + payload);
        if (resp.duration.count() >= 4500 &&
            baseline.duration.count() < 2000) {
          findings.push_back({"SQLi (Time)", "high", url,
                              "Time-based blind SQLi: " +
                                  std::to_string(resp.duration.count()) + "ms",
                              param, payload, ""});
          break;
        }
      }
    }
  }
  return findings;
}

} // namespace

std::vector<Scanner> register_smart_scanners() {
  return {
      {"SQLi (Blind Boolean)", scan_sqli_blind_boolean},
      {"XSS (Context-Aware)", scan_xss_context_aware},
      {"WAF Detection", scan_waf_detect},
      {"SQLi (Time-Based)", scan_sqli_time_based},
  };
}

} // namespace apex

File: pkg/scanner/smart.go
package scanner

import (
	"crypto/md5"
	"fmt"
	"net/url"
	"regexp"
	"strings"
	"sync"

	"github.com/zanicool/apex-cli/pkg/crawler"
	"github.com/zanicool/apex-cli/pkg/engine"
	"github.com/zanicool/apex-cli/pkg/oob"
)

// --- Boolean-Based Blind SQLi ---

func scanSQLiBlindBoolean(cfg *engine.Config, http *engine.HTTPClient, crawl *crawler.Result, _ *oob.Client) []Finding {
	var findings []Finding
	var mu sync.Mutex
	sem := make(chan struct{}, cfg.Threads)
	var wg sync.WaitGroup

	trueFalse := []struct{ truePayload, falsePayload string }{
		{"' AND '1'='1", "' AND '1'='2"},
		{"' AND 1=1--", "' AND 1=2--"},
		{"\" AND \"1\"=\"1", "\" AND \"1\"=\"2"},
		{" AND 1=1", " AND 1=2"},
		{"') AND ('1'='1", "') AND ('1'='2"},
		{" OR 1=1--", " OR 1=2--"},
	}

	for u, params := range crawl.Params {
		for _, p := range params {
			wg.Add(1)
			go func(baseURL, param string) {
				defer wg.Done()
				sem <- struct{}{}
				defer func() { <-sem }()

				// Get baseline
				baseResp := http.Get(baseURL)
				if baseResp.Err != nil {
					return
				}
				baseHash := hashBody(baseResp.Body)
				baseSize := len(baseResp.Body)

				for _, tf := range trueFalse {
					trueURL := injectParam(baseURL, param, tf.truePayload)
					falseURL := injectParam(baseURL, param, tf.falsePayload)

					trueResp := http.Get(trueURL)
					falseResp := http.Get(falseURL)

					if trueResp.Err != nil || falseResp.Err != nil {
						continue
					}

					trueHash := hashBody(trueResp.Body)
					falseHash := hashBody(falseResp.Body)

					// Boolean blind: true condition matches baseline, false differs
					if trueHash == baseHash && falseHash != baseHash {
						mu.Lock()
						findings = append(findings, Finding{
							Type: "SQL Injection (Boolean-Based Blind)", Severity: "critical",
							URL: trueURL, Param: param,
							Payload: fmt.Sprintf("TRUE: %s | FALSE: %s", tf.truePayload, tf.falsePayload),
							Detail:  fmt.Sprintf("True response matches baseline (%d bytes), false differs (%d bytes)", baseSize, len(falseResp.Body)),
							Template: "apex-sqli-boolean",
						})
						mu.Unlock()
						return
					}

					// Also check: true differs from false (both differ from baseline but differently)
					trueSize := len(trueResp.Body)
					falseSize := len(falseResp.Body)
					if trueHash != falseHash && abs(trueSize-baseSize) < 50 && abs(falseSize-baseSize) > 100 {
						mu.Lock()
						findings = append(findings, Finding{
							Type: "SQL Injection (Boolean-Based Blind)", Severity: "critical",
							URL: trueURL, Param: param,
							Payload: fmt.Sprintf("TRUE: %s | FALSE: %s", tf.truePayload, tf.falsePayload),
							Detail:  fmt.Sprintf("True=%d bytes (≈baseline), False=%d bytes (differs)", trueSize, falseSize),
							Template: "apex-sqli-boolean",
						})
						mu.Unlock()
						return
					}
				}
			}(u, p)
		}
	}
	wg.Wait()
	return findings
}

// --- Context-Aware XSS ---

func scanXSSContextAware(cfg *engine.Config, http *engine.HTTPClient, crawl *crawler.Result, _ *oob.Client) []Finding {
	var findings []Finding
	var mu sync.Mutex
	sem := make(chan struct{}, cfg.Threads)
	var wg sync.WaitGroup

	canary := "apex7x7z"

	for u, params := range crawl.Params {
		for _, p := range params {
			wg.Add(1)
			go func(baseURL, param string) {
				defer wg.Done()
				sem <- struct{}{}
				defer func() { <-sem }()

				// Inject canary to find reflection context
				canaryURL := injectParam(baseURL, param, canary)
				resp := http.Get(canaryURL)
				if resp.Err != nil || !strings.Contains(resp.Body, canary) {
					return // Not reflected
				}

				// Determine context
				ctx := detectContext(resp.Body, canary)
				payloads := getContextPayloads(ctx)

				for _, payload := range payloads {
					testURL := injectParam(baseURL, param, payload.value)
					testResp := http.Get(testURL)
					if testResp.Err != nil {
						continue
					}
					if strings.Contains(testResp.Body, payload.detect) {
						mu.Lock()
						findings = append(findings, Finding{
							Type: fmt.Sprintf("XSS (%s context)", ctx), Severity: "high",
							URL: testURL, Param: param, Payload: payload.value,
							Evidence: payload.detect,
							Detail:   fmt.Sprintf("Reflected in %s context — payload breaks out and executes", ctx),
							Template: "apex-xss-context",
						})
						mu.Unlock()
						return
					}
				}
			}(u, p)
		}
	}
	wg.Wait()
	return findings
}

type xssPayload struct {
	value  string
	detect string
}

func detectContext(body, canary string) string {
	idx := strings.Index(body, canary)
	if idx < 0 {
		return "html"
	}
	// Look at surrounding characters
	before := ""
	if idx > 50 {
		before = body[idx-50 : idx]
	} else if idx > 0 {
		before = body[:idx]
	}

	if strings.Contains(before, "<script") || strings.Contains(before, "var ") || strings.Contains(before, "= '") || strings.Contains(before, "= \"") {
		if strings.Contains(before, "'") {
			return "js-single"
		}
		return "js-double"
	}
	if regexp.MustCompile(`\w+\s*=\s*"[^"]*$`).MatchString(before) {
		return "attr-double"
	}
	if regexp.MustCompile(`\w+\s*=\s*'[^']*$`).MatchString(before) {
		return "attr-single"
	}
	if strings.Contains(before, "href=") || strings.Contains(before, "src=") {
		return "url"
	}
	return "html"
}

func getContextPayloads(ctx string) []xssPayload {
	switch ctx {
	case "js-single":
		return []xssPayload{
			{"'-alert(1)-'", "-alert(1)-"},
			{"';alert(1)//", ";alert(1)"},
			{"\\'-alert(1)//", "-alert(1)"},
		}
	case "js-double":
		return []xssPayload{
			{"\"-alert(1)-\"", "-alert(1)-"},
			{"\";alert(1)//", ";alert(1)"},
		}
	case "attr-double":
		return []xssPayload{
			{"\" onmouseover=alert(1) \"", "onmouseover=alert(1)"},
			{"\" onfocus=alert(1) autofocus=\"", "onfocus=alert(1)"},
			{"\"><script>alert(1)</script>", "<script>alert(1)</script>"},
			{"\"><img src=x onerror=alert(1)>", "onerror=alert(1)"},
		}
	case "attr-single":
		return []xssPayload{
			{"' onmouseover=alert(1) '", "onmouseover=alert(1)"},
			{"' onfocus=alert(1) autofocus='", "onfocus=alert(1)"},
			{"'><img src=x onerror=alert(1)>", "onerror=alert(1)"},
		}
	case "url":
		return []xssPayload{
			{"javascript:alert(1)", "javascript:alert(1)"},
			{"data:text/html,<script>alert(1)</script>", "data:text/html"},
		}
	default: // html
		return []xssPayload{
			{"<script>alert(1)</script>", "<script>alert(1)</script>"},
			{"<img src=x onerror=alert(1)>", "onerror=alert(1)"},
			{"<svg onload=alert(1)>", "onload=alert(1)"},
			{"<details open ontoggle=alert(1)>", "ontoggle=alert(1)"},
		}
	}
}

// --- WAF Detection (run before scanning to adapt strategy) ---

func DetectWAF(http *engine.HTTPClient, targets []string) []string {
	var wafs []string
	if len(targets) == 0 {
		return wafs
	}

	// Send obvious attack payload and check response
	testURL := targets[0] + "/?test=<script>alert(1)</script>&id=1' OR 1=1--"
	resp := http.Get(testURL)
	if resp.Err != nil {
		return wafs
	}

	wafSigs := map[string][]string{
		"Cloudflare":  {"cf-ray", "cloudflare", "__cfduid"},
		"AWS WAF":     {"awselb", "x-amzn-requestid", "aws"},
		"Akamai":      {"akamai", "x-akamai"},
		"Imperva":     {"incapsula", "imperva", "x-iinfo"},
		"F5 BIG-IP":   {"bigip", "f5", "ts="},
		"Sucuri":      {"sucuri", "x-sucuri"},
		"ModSecurity": {"mod_security", "modsecurity"},
		"Fortinet":    {"fortigate", "fortiweb"},
	}

	headerStr := strings.ToLower(fmt.Sprintf("%v", resp.Headers))
	bodyLower := strings.ToLower(resp.Body[:min(2000, len(resp.Body))])
	combined := headerStr + bodyLower

	for name, sigs := range wafSigs {
		for _, sig := range sigs {
			if strings.Contains(combined, sig) {
				wafs = append(wafs, name)
				break
			}
		}
	}

	// Also detect by status code pattern
	if resp.StatusCode == 403 || resp.StatusCode == 406 || resp.StatusCode == 429 {
		if len(wafs) == 0 {
			wafs = append(wafs, "Unknown WAF")
		}
	}

	return wafs
}

// --- Smart Deduplication ---

func DeduplicateFindings(findings []Finding) []Finding {
	seen := make(map[string]bool)
	var result []Finding

	for _, f := range findings {
		// Deduplicate by type + base URL (ignore query params for dedup)
		baseURL := f.URL
		if parsed, err := url.Parse(f.URL); err == nil {
			baseURL = parsed.Scheme + "://" + parsed.Host + parsed.Path
		}
		key := f.Type + "|" + baseURL + "|" + f.Param

		if !seen[key] {
			seen[key] = true
			result = append(result, f)
		}
	}
	return result
}

// --- Wildcard Detection ---

func IsWildcard(http *engine.HTTPClient, target string) bool {
	// Request 3 random non-existent paths
	randoms := []string{"/apex_random_8f3k2j", "/apex_nonexist_9x7m4p", "/apex_test_2w5n8q"}
	var hashes []string

	for _, path := range randoms {
		resp := http.Get(target + path)
		if resp.Err != nil || resp.StatusCode == 404 {
			return false // Proper 404 = not wildcard
		}
		if resp.StatusCode == 200 {
			hashes = append(hashes, hashBody(resp.Body))
		}
	}

	// If all 3 random paths return same content = wildcard
	if len(hashes) == 3 && hashes[0] == hashes[1] && hashes[1] == hashes[2] {
		return true
	}
	return false
}

// --- Helpers ---

func hashBody(body string) string {
	// Normalize: strip whitespace-only differences
	normalized := strings.TrimSpace(body)
	h := md5.Sum([]byte(normalized))
	return fmt.Sprintf("%x", h)
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

// IsCloudflareChallenge checks if a target returns a Cloudflare challenge/block page
func IsCloudflareChallenge(http *engine.HTTPClient, target string) bool {
	resp := http.Get(target)
	if resp.Err != nil {
		return false
	}
	body := resp.Body
	// Cloudflare managed challenge markers
	cfMarkers := []string{
		"challenges.cloudflare.com",
		"cf-chl-bypass",
		"_cf_chl_opt",
		"Just a moment...",
		"cf-error-details",
		"cdn-cgi/challenge-platform",
	}
	for _, marker := range cfMarkers {
		if strings.Contains(body, marker) {
			return true
		}
	}
	return false
}

// IsSPACatchAll checks if a target returns identical responses for any path (SPA with client-side routing)
func IsSPACatchAll(http *engine.HTTPClient, target string) bool {
	paths := []string{"/", "/apex_fake_path_x9k2m", "/apex_fake_path_q7w3n", "/checkout/complete", "/api/upload"}
	var bodies []string
	for _, p := range paths {
		resp := http.Get(target + p)
		if resp.Err != nil {
			return false
		}
		if resp.StatusCode == 404 || resp.StatusCode == 403 || resp.StatusCode == 301 || resp.StatusCode == 302 {
			return false
		}
		bodies = append(bodies, hashBody(resp.Body))
	}
	// If all paths return identical content, it's a SPA catch-all
	if len(bodies) >= 4 {
		for i := 1; i < len(bodies); i++ {
			if bodies[i] != bodies[0] {
				return false
			}
		}
		return true
	}
	return false
}
